content/manuals/build/building/cdi.md

```diff
@@ -0,0 +1,531 @@
+
+```text
+Devices:
+ Name:                  vendor1.com/device=foo
+ Automatically allowed: false
+```
+
+To allow the device, you can use the [`--allow` flag](/reference/cli/docker/buildx/build.md#allow)
+with the `docker buildx build` command:
+
+```console
+$ docker buildx build --allow device .
+```
+
+Or you can set the `org.mobyproject.buildkit.device.autoallow` annotation in
+the CDI specification to automatically allow the device for all builds:
+
+```yaml {title="/etc/cdi/foo.yaml"}
+cdiVersion: "0.6.0"
+kind: "vendor1.com/device"
+devices:
+- name: foo
+  containerEdits:
+    env:
+    - FOO=injected
+annotations:
+  org.mobyproject.buildkit.device.autoallow: true
+```
+
+Now running the build again with the `--allow device` flag:
+
+```console
+$ docker buildx build --progress=plain --allow device .
+#0 building with "default" instance using docker driver
+
+#1 [internal] load build definition from Dockerfile
+#1 transferring dockerfile: 159B done
+#1 DONE 0.0s
+
+#2 resolve image config for docker-image://docker.io/docker/dockerfile-upstream:1.17-labs
+#2 DONE 0.1s
+
+#3 docker-image://docker.io/docker/dockerfile-upstream:1.17-labs@sha256:9187104f31e3a002a8a6a3209ea1f937fb7486c093cbbde1e14b0fa0d7e4f1b5
+#3 CACHED
+
+#4 [internal] load metadata for docker.io/library/busybox:latest
+#4 DONE 0.1s
+
+#5 [internal] load .dockerignore
+#5 transferring context: 2B done
+#5 DONE 0.0s
+
+#6 [1/2] FROM docker.io/library/busybox:latest@sha256:f85340bf132ae937d2c2a763b8335c9bab35d6e8293f70f606b9c6178d84f42b
+#6 CACHED
+
+#7 [2/2] RUN --device=vendor1.com/device   env | grep ^FOO=
+#7 0.155 FOO=injected
+#7 DONE 0.2s
+```
+
+The build is successful and the output shows that the `FOO` environment variable
+was injected into the build environment as specified in the CDI specification.
+
+## Set up a container builder with GPU support
+
+In this section, we will show you how to set up a [container builder](../builders/drivers/docker-container.md)
+using NVIDIA GPUs. Since Buildx v0.22, when creating a new container builder, a
+GPU request is automatically added to the container builder if the host has GPU
+drivers installed in the kernel. This is similar to using [`--gpus=all` with the `docker run`](/reference/cli/docker/container/run.md#gpus)
+command.
+
+> [!NOTE]
+> We made a specially crafted BuildKit image because the current BuildKit
+> release image is based on Alpine that doesn’t support NVIDIA drivers. The
+> following image is based on Ubuntu and will install the NVIDIA client
+> libraries and generate the CDI specification for your GPU in the container
+> builder if a device is requested during a build. This image is temporarily
+> hosted on Docker Hub under `crazymax/buildkit:v0.23.2-ubuntu-nvidia`.
+
+Now let's create a container builder named `gpubuilder` using Buildx:
+
+```console
+$ docker buildx create --name gpubuilder --driver-opt "image=crazymax/buildkit:v0.23.2-ubuntu-nvidia" --bootstrap
```

Comment:
We probably want to have this somewhere more "official" I guess (not a blocker); how complicated is this image to build? Would it be an option to add the `Dockerfile` here (if it's just 2 .. 3 steps)?

Reply:
Yes it will be in the future, this is tracked in https://github.com/moby/buildkit/pull/5944.

> Would it be an option to add the `Dockerfile` here (if it's just 2 .. 3 steps)?

It's the BuildKit Dockerfile so not an option :sweat_smile: